aspnetcore/performance/response-compression.md

```diff
@@ -0,0 +1,167 @@
+
+Submit a request to the sample application without the `Accept-Encoding` header and observe that the response is uncompressed. The `Content-Encoding` and `Vary` headers are not present on the response.
+
+![Fiddler window showing result of a request without the Accept-Encoding header. The response is not compressed.](response-compression/_static/request-uncompressed.png)
+
+Submit a request to the sample application with the `Accept-Encoding: gzip` header and observe that the response is compressed. The `Content-Encoding` and `Vary` headers are present on the response.
+
+![Fiddler window showing result of a request with the Accept-Encoding header and a value of gzip. The Vary and Content-Encoding headers are added to the response. The response is compressed.](response-compression/_static/request-compressed.png)
+
+### Providers
+#### GzipCompressionProvider
+Use the `GzipCompressionProvider` to compress responses with GZip. This is the default compression provider if none are specified. You can set the compression level with the `GzipCompressionProviderOptions`. The default is `CompressionLevel.Fastest`.
+
+Compression Level | Description
+--- | ---
+`CompressionLevel.Fastest` | Compression should complete as quickly as possible, even if the resulting output is not optimally compressed.
+`CompressionLevel.NoCompression` | No compression should be performed.
+`CompressionLevel.Optimal` | Responses should be optimally compressed, even if the compression takes more time to complete.
+
+[!code-csharp[Main](response-compression/sample/Startup.cs?name=snippet2&highlight=3)]
+
+[!code-csharp[Main](response-compression/sample/Startup.cs?name=snippet3)]
+
+#### Custom Providers
+You can create custom compression implementations with `ICompressionProvider`. The `EncodingName` represents the content encoding that this `ICompressionProvider` produces. The middleware will use this information to choose the provider based on the list specified in the `Accept-Encoding` header of the request.
+
+Using the sample application, the client would submit a request with the `Accept-Encoding: custom` header. The middleware will use the custom compression implementation and return the response with a `Content-Encoding: custom` header. The client must be able to decompress the custom encoding in order for a custom compression implementation to work.
+
+[!code-csharp[Main](response-compression/sample/Startup.cs?name=snippet2&highlight=4)]
+
+[!code-csharp[Main](response-compression/sample/CustomCompressionProvider.cs?name=snippet1)]
+
+Submit a request to the sample application with the `Accept-Encoding: custom` header and observe the response headers. The `Vary` and `Content-Encoding` headers are present on the response. The response body (not shown) isn't compressed by the sample, as there is no compression implementation in the `CustomCompressionProvider` class of the sample. However, the sample shows where you would implement such a compression algorithm.
```

Comment:
don't use `custom`, use something that should more obviously be renamed like `mycustomcompression`.